File: 0207-course-schedule/0207-course-schedule.py
class Solution:
    def canFinish(self, numCourses: int, prerequisites: List[List[int]]) -> bool:
        queue = deque()
        adjList = [0 for i in range(numCourses)]
        hashMap = dict() 
        count = 0
        for course in prerequisites:
            adjList[course[0]] += 1

            if course[1] not in hashMap.keys():
                hashMap[course[1]] = []
            
            hashMap.get(course[1]).append(course[0])
        
        for i in range(len(adjList)):
            if adjList[i] == 0:
                queue.append(i)
                count += 1

        if len(queue) == 0:
            return False

        while len(queue) > 0:
            curr = queue.popleft()
            currList = hashMap.get(curr)
            # print(curr)
            if currList != None:
                for child in currList:
                    adjList[child] -= 1

                    if adjList[child] == 0:
                        queue.append(child)
                        count += 1
        
        if count == numCourses:
            return True
        
        return False

File: 0207-course-schedule/0207-course-schedule.java
class Solution {
    public boolean canFinish(int numCourses, int[][] prerequisites) {
        int[] indegrees = new int[numCourses];
        HashMap<Integer, List<Integer>> map = new HashMap<>();

        int count = 0;

        for(int [] course: prerequisites){
            indegrees[course[0]]++;

            if(!map.containsKey(course[1])){
                map.put(course[1], new ArrayList<>());
            }

            map.get(course[1]).add(course[0]);
        }

        Queue<Integer> q = new LinkedList<>();

        for(int i = 0; i < indegrees.length; i++){
            if(indegrees[i] == 0){
                q.add(i);
                count++;
            }
        }

        if(q.isEmpty()){
            return false;
        }

        while(!q.isEmpty()){
            int curr = q.poll();
            List<Integer> currList = map.get(curr);

            if(currList != null){
                for(int child : currList){
                    indegrees[child]--;

                    if(indegrees[child] == 0){
                        q.add(child);
                        count++;
                    }
                }
            }
        }


        return count == numCourses;
    }
}

//Graph
// Time Complexity: O(V+E)
// Space Complexity: O(V+E)